File: java/src/test/java/it/giannibombelli/wsc2026/booking/application/integration/payment/adapter/PaymentResultTest.java
package it.giannibombelli.wsc2026.booking.application.integration.payment.adapter;

import it.giannibombelli.wsc2026.booking.application.commands.BookingConfirmationCommands;
import it.giannibombelli.wsc2026.booking.domain.booking.Booking;
import it.giannibombelli.wsc2026.booking.domain.booking.BookingId;
import it.giannibombelli.wsc2026.booking.domain.ports.BookingRepository;
import it.giannibombelli.wsc2026.booking.infrastructure.SqliteBookingRepository;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.payment.integration.PaymentResultIntegrationEvent;
import it.giannibombelli.wsc2026.testsupport.AggregateFactory;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import javax.sql.DataSource;
import java.math.BigDecimal;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class PaymentResultTest {

    private BookingRepository repository;
    private PaymentResult paymentResult;

    @BeforeEach
    void setUp() {
        DataSource dataSource = DatabaseSetup.initializeInMemoryDb("booking");
        repository = new SqliteBookingRepository(dataSource);
        paymentResult = new PaymentResult(repository);
    }

    @Test
    void adaptPaymentAccepted_returnsConfirmBooking() {
        Booking booking = saveBooking();
        Money amount = new Money(new BigDecimal("50.00"));
        PaymentResultIntegrationEvent event = new PaymentResultIntegrationEvent.PaymentAcceptedIntegrationEvent(
            booking.id().value().toString(), amount
        );

        BookingConfirmationCommands cmd = paymentResult.adapt(event);

        assertThat(cmd).isNotNull();
        assertThat(cmd).isInstanceOf(BookingConfirmationCommands.ConfirmBooking.class);
        assertThat(cmd.aggregateId()).isEqualTo(booking.id());
        assertThat(((BookingConfirmationCommands.ConfirmBooking) cmd).giftCardReference()).isEqualTo(booking.giftCardReference());
        assertThat(((BookingConfirmationCommands.ConfirmBooking) cmd).amount()).isEqualTo(amount);
    }

    @Test
    void adaptPaymentRejected_returnsRejectBooking() {
        Booking booking = saveBooking();
        Money amount = new Money(new BigDecimal("50.00"));
        PaymentResultIntegrationEvent event = new PaymentResultIntegrationEvent.PaymentRejectedIntegrationEvent(
            booking.id().value().toString(), amount, "declined"
        );

        BookingConfirmationCommands cmd = paymentResult.adapt(event);

        assertThat(cmd).isNotNull();
        assertThat(cmd).isInstanceOf(BookingConfirmationCommands.RejectBooking.class);
        assertThat(cmd.aggregateId()).isEqualTo(booking.id());
        assertThat(((BookingConfirmationCommands.RejectBooking) cmd).giftCardReference()).isEqualTo(booking.giftCardReference());
        assertThat(((BookingConfirmationCommands.RejectBooking) cmd).amount()).isEqualTo(amount);
    }

    @Test
    void adaptPaymentExpired_returnsNull() {
        Booking booking = saveBooking();
        PaymentResultIntegrationEvent event = new PaymentResultIntegrationEvent.PaymentExpiredIntegrationEvent(
            booking.id().value().toString(), new Money(new BigDecimal("50.00"))
        );

        BookingConfirmationCommands cmd = paymentResult.adapt(event);

        assertThat(cmd).isNull();
    }

    @Test
    void adaptUnknownBooking_returnsNull() {
        PaymentResultIntegrationEvent event = new PaymentResultIntegrationEvent.PaymentAcceptedIntegrationEvent(
            UUID.randomUUID().toString(), new Money(new BigDecimal("50.00"))
        );

        BookingConfirmationCommands cmd = paymentResult.adapt(event);

        assertThat(cmd).isNull();
    }

    @Test
    void adaptNullEvent_throwsException() {
        assertThatThrownBy(() -> paymentResult.adapt(null))
            .isInstanceOf(IllegalArgumentException.class);
    }

    private Booking saveBooking() {
        Booking booking = AggregateFactory.createBooking();
        repository.save(booking);
        return booking;
    }
}

File: typescript/tests/booking/application/integration/payment/adapter/paymentResult.test.ts
import { describe, expect, it } from 'vitest';
import { Money } from '@/common/domain/primitive/money.js';
import {
  paymentAcceptedIntegrationEvent,
  paymentExpiredIntegrationEvent,
  paymentRejectedIntegrationEvent,
} from '@/payment/integration/paymentResultIntegrationEvent.js';
import { PaymentResult } from '@/booking/application/integration/payment/adapter/paymentResult.js';
import { SqliteBookingRepository } from '@/booking/infrastructure/sqliteBookingRepository.js';
import { DatabaseSetup } from '../../../../../testsupport/databaseSetup.js';
import { BookingAggregateFactory } from '../../../../../testsupport/booking/aggregateFactory.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';

describe('PaymentResult', () => {
  const database = DatabaseSetup.initializeInMemoryDb('booking');
  const repository = new SqliteBookingRepository(database);
  const paymentResult = new PaymentResult(repository);

  function saveBooking() {
    const booking = BookingAggregateFactory.createBooking();
    repository.save(booking);
    return booking;
  }

  it('adapts PaymentAccepted to ConfirmBooking', () => {
    const booking = saveBooking();
    const amount = new Money(50);
    const event = paymentAcceptedIntegrationEvent(booking.id().value.value, amount);

    const result = paymentResult.adapt(event);

    expect(result).not.toBeNull();
    expect(result!.kind).toBe('ConfirmBooking');
    expect(result!.aggregateId).toEqual(booking.id());
    expect(result!.giftCardReference).toEqual(booking.giftCardReference());
    expect(result!.amount).toEqual(amount);
  });

  it('adapts PaymentRejected to RejectBooking', () => {
    const booking = saveBooking();
    const amount = new Money(50);
    const event = paymentRejectedIntegrationEvent(booking.id().value.value, amount, 'declined');

    const result = paymentResult.adapt(event);

    expect(result).not.toBeNull();
    expect(result!.kind).toBe('RejectBooking');
    expect(result!.aggregateId).toEqual(booking.id());
    expect(result!.giftCardReference).toEqual(booking.giftCardReference());
    expect(result!.amount).toEqual(amount);
  });

  it('adapts PaymentExpired to null', () => {
    const booking = saveBooking();
    const event = paymentExpiredIntegrationEvent(booking.id().value.value, new Money(50));

    const result = paymentResult.adapt(event);

    expect(result).toBeNull();
  });

  it('adapts unknown booking to null', () => {
    const event = paymentAcceptedIntegrationEvent(Uuid.generate().value, new Money(50));

    const result = paymentResult.adapt(event);

    expect(result).toBeNull();
  });

  it('throws on null event', () => {
    expect(() => paymentResult.adapt(null as unknown as Parameters<typeof paymentResult.adapt>[0])).toThrow();
  });
});
